### PR TITLE
OpenAPI: Deprecate `oauth/tokens` endpoint

### PR DESCRIPTION
**Pull request by snazy**
This PR implements "M1" of [this document](https://docs.google.com/document/d/1Xi5MRk8WdBWFC3N_eSmVcrLhk3yu5nJ9x_wC0ec6kVQ/), see #10537.

### CODE REVIEW

File: open-api/rest-catalog-open-api.yaml [hunk 1]
```diff
@@ -134,9 +134,19 @@ paths:
     post:
       tags:
         - OAuth2 API
-      summary: Get a token using an OAuth2 flow
+      summary: Get a token using an OAuth2 flow (DEPRECATED for REMOVAL)
       operationId: getToken
       description:
+        The `oauth/tokens` endpoint is **DEPRECATED for REMOVAL**. It is not recommended to _not_
+        implement this endpoint, unless you are fully aware of the potential security implications.
```
[the remark this answers is not in the document]
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml]: Thanks - fixed

**Review by dimas-b — APPROVED** [review]

File: core/src/test/resources/logback-test.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<!--
+
+    Copyright (C) 2020 Dremio
```
Comment (nastra) [inline core/src/test/resources/logback-test.xml:4]: why is this file needed?

File: core/src/test/resources/logback-test.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  <contextListener class="ch.qos.logback.classic.jul.LevelChangePropagator"/>
+  <appender name="console" class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%date{ISO8601} [%thread] %-5level %logger{36} - %msg%n</pattern>
+    </encoder>
+  </appender>
+  <root level="${test.log.level:-WARN}">
+    <appender-ref ref="console"/>
+  </root>
+  <logger name="org.projectnessie.client.auth.oauth2" level="INFO"/>
```
Comment (nastra) [inline core/src/test/resources/logback-test.xml:29]: references stuff not used in Iceberg

File: core/src/test/resources/logback-test.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<configuration debug="true">
```
Comment (snazy) [inline core/src/test/resources/logback-test.xml:22]: This file is used to inspect the logged messages.

File: open-api/rest-catalog-open-api.yaml [hunk 5]
```diff
@@ -134,9 +134,19 @@ paths:
     post:
       tags:
         - OAuth2 API
-      summary: Get a token using an OAuth2 flow
+      summary: Get a token using an OAuth2 flow (DEPRECATED for REMOVAL)
```
Comment (ajantha-bhat) [inline open-api/rest-catalog-open-api.yaml:137]: We should add the field  `deprecated: true` to mark it as deprecated.

 ```
 /v1/oauth/tokens:

    post:
      tags:
        - OAuth2 API
      summary: Get a token using an OAuth2 flow
      deprecated: true
      operationId: getToken
      description: ---
```

Comment (ajantha-bhat) [reply open-api/rest-catalog-open-api.yaml:137]: more info: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#fixed-fields-8

**Review by ajantha-bhat — APPROVED** [review]

File: core/src/test/resources/logback-test.xml (hunk 4 above)
Comment (snazy) [reply core/src/test/resources/logback-test.xml:22]: It's required to at least manually verify that the change emits the expected log message.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 6]
```diff
@@ -183,6 +183,17 @@ public void initialize(String name, Map<String, String> unresolved) {
     String credential = props.get(OAuth2Properties.CREDENTIAL);
     String scope = props.getOrDefault(OAuth2Properties.SCOPE, OAuth2Properties.CATALOG_SCOPE);
     Map<String, String> optionalOAuthParams = OAuth2Util.buildOptionalParam(props);
+    if (!props.containsKey(OAuth2Properties.OAUTH2_SERVER_URI)) {
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:186]: This would be a problem for sigv4 users. The Sigv4 code path is currently hidden in the HTTPClient with:

```
private static final String SIGV4_ENABLED = "rest.sigv4-enabled";
  private static final String SIGV4_REQUEST_INTERCEPTOR_IMPL =
      "org.apache.iceberg.aws.RESTSigV4Signer";
```

So we should at least check this property and not log warning if the flag is set to true.

I can do some refactoring to bring them to public after this PR is merged, by formally introduce an additional property like `auth-mode`, that can be of values like `oauth2`, `amazon-sigv4`, and defaults to `oauth2`. And we only check this property if the auth mode is oauth2 or null.

File: core/src/test/resources/logback-test.xml (hunk 4 above)
Comment (jackye1995) [reply core/src/test/resources/logback-test.xml:22]: I am definitely for adding a generic logging file for tests, but this log file seems too specific to testing this warning, I would suggest we add it separately.

Comment (jackye1995) [reply core/src/test/resources/logback-test.xml:22]: The use of which logging system is also subject to some discussions

File: open-api/rest-catalog-open-api.yaml [hunk 7]
```diff
@@ -2922,7 +2933,10 @@ components:
         See https://datatracker.ietf.org/doc/html/rfc8693#section-3
 
     OAuthClientCredentialsRequest:
-      description:
+      deprecated: true
+      description: The `oauth/tokens` endpoint and related schemas are **DEPRECATED for REMOVAL** from this spec.
```
Comment (jackye1995) [inline open-api/rest-catalog-open-api.yaml:2937]:
```suggestion
      description: 
      
        The `oauth/tokens` endpoint and related schemas are **DEPRECATED for REMOVAL** from this spec.
```

Comment (jackye1995) [reply open-api/rest-catalog-open-api.yaml:2937]: And similar comments for other descriptions

File: open-api/rest-catalog-open-api.yaml [hunk 8]
```diff
@@ -134,9 +134,20 @@ paths:
...
-      summary: Get a token using an OAuth2 flow
+      summary: Get a token using an OAuth2 flow (DEPRECATED for REMOVAL)
+      deprecated: true
       operationId: getToken
       description:
+        The `oauth/tokens` endpoint is **DEPRECATED for REMOVAL**. It is not recommended to _not_
+        implement this endpoint, unless you are fully aware of the potential security implications.
+        All clients are encouraged to explicitly set the configuration property `oauth2-server-uri`
+        to the correct OAuth endpoint.
+        See [Security improvements in the Iceberg REST specification](https://github.com/apache/iceberg/issues/10537)
```
Comment (jackye1995) [inline open-api/rest-catalog-open-api.yaml:145]: I think instead of linking to the GitHub issue, we probably want to add a dedicated section explaining OAuth2 usage, maybe in `open-api/security/oauth2.md`. Should we do it as a part of this PR? I am okay either way.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 6 above)
Comment (snazy) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:186]: Good point

Comment (snazy) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:186]: But we can do the `auth-mode` change in the later "milestone"

Comment (snazy) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java]: @jackye1995 can you take a look at the updated `if` condition?

File: open-api/rest-catalog-open-api.yaml (hunk 8 above)
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:145]: Not sure TBH. Since the whole oauth stuff is eventually removed from the spec, I think it's okay to leave the link to the issue.

File: open-api/rest-catalog-open-api.yaml [hunk 9]
```diff
@@ -134,9 +134,20 @@ paths:
     post:
       tags:
         - OAuth2 API
-      summary: Get a token using an OAuth2 flow
+      summary: Get a token using an OAuth2 flow (DEPRECATED for REMOVAL)
+      deprecated: true
       operationId: getToken
       description:
+        The `oauth/tokens` endpoint is **DEPRECATED for REMOVAL**. It is not recommended to _not_
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:141]: `It is not recommended to _not_ implement this endpoint` seems quite confusing. Did you mean `It is not recommended to implement this endpoint...`?

File: open-api/rest-catalog-open-api.yaml [hunk 10]
```diff
@@ -134,9 +134,20 @@ paths:
...
+        implement this endpoint, unless you are fully aware of the potential security implications.
+        All clients are encouraged to explicitly set the configuration property `oauth2-server-uri`
+        to the correct OAuth endpoint.
+        See [Security improvements in the Iceberg REST specification](https://github.com/apache/iceberg/issues/10537)
+
+
+        Old description follows
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:148]:
```suggestion
```

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 11]
```diff
@@ -176,19 +176,34 @@ public void initialize(String name, Map<String, String> unresolved) {
...
+    boolean hasInitToken = initToken != null;
 
     // fetch auth and config to complete initialization
     ConfigResponse config;
     OAuthTokenResponse authResponse;
     String credential = props.get(OAuth2Properties.CREDENTIAL);
+    boolean hasCredential = credential != null && !credential.isEmpty();
     String scope = props.getOrDefault(OAuth2Properties.SCOPE, OAuth2Properties.CATALOG_SCOPE);
     Map<String, String> optionalOAuthParams = OAuth2Util.buildOptionalParam(props);
+    if (!props.containsKey(OAuth2Properties.OAUTH2_SERVER_URI)
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: `initialize` is already quite long and the if condition is complicated here, so might be worth extracting this into a separate method

File: open-api/rest-catalog-open-api.yaml (hunk 10 above)
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:148]: Why??

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 11 above)
Comment (snazy) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: Is this _really_ important or rather a nit for code that is going to be changes in the near future?

File: open-api/rest-catalog-open-api.yaml (hunk 10 above)
Comment (nastra) [reply open-api/rest-catalog-open-api.yaml:148]: it doesn't improve this section and rather adds confusion when reading the entire section, hence I'd just remove it

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 11 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: yeah I think if we want to formally refactor this, it might be okay to leave the warning here for now. What do you think Eduard?

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: I'm ok leaving it as-is but it adds complexity to the already difficult to understand `initialize` method, hence why I suggested to consider moving this into a separate method (even if the code eventually gets changed in the near future we don't necessarily have to add additional complexity if we can avoid it in the first place)

Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: I am doing a prototype for the refactoring, maybe let me publish a draft and we can see how that would look like

Comment (snazy) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:188]: I'd rather let this change not depend on a follow-up, but get it in.

**Review by jackye1995 — APPROVED** [review]
Looks good to me

File: open-api/rest-catalog-open-api.yaml [hunk 12]
```diff
@@ -134,9 +134,17 @@ paths:
     post:
       tags:
         - OAuth2 API
-      summary: Get a token using an OAuth2 flow
+      summary: Get a token using an OAuth2 flow (DEPRECATED for REMOVAL)
+      deprecated: true
       operationId: getToken
       description:
+        The `oauth/tokens` endpoint is **DEPRECATED for REMOVAL**. It is _not_ recommended to
```
[the remark this answers is not in the document]
Comment (jackye1995) [reply open-api/rest-catalog-open-api.yaml:141]: What does target version mean here? The spec does not really have a general version the last time we discussed.

[the remark this answers is not in the document]
Comment (danielcweeks) [reply open-api/rest-catalog-open-api.yaml:141]: I think the target version should be tied to the reference implementation where the default behaviors will change.

As we discussed before, we can change the requirement that URI be set with the 1.7 release.  However, we can't change the behaviors until 2.0.  So I think we should reference the 2.0 release as the target for removal.

In general, we shouldn't add deprecations that don't clearly indicate a target for removal and reference to the replacement.  In this case we don't have clarity on the exact replacement but I think the current reference to the issue is fine.

[the remark this answers is not in the document]
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:141]: I'm proposing to include it in 1.6.0 and not wait for 1.7.0 - added a note that mentions "Iceberg (Java) **1.6.0**", because that's the next release this change can realistically go into.

Sure, the improvement proposal document mentioned Iceberg 1.7.0, because at the time of writing it (June), 1.7.0 was the next release the deprecation could go in. 1.6.0 was considered to be released in June ([meeting notes](https://docs.google.com/document/d/1YuGhUdukLP5gGiqCbk0A5_Wifqe2CZWgOd3TbhY3UQg/edit#heading=h.z3dncl7gr8m1)). But now it's already July and 1.6.0 is still not ready to be released.

This change is part of an effort to address a potential security issue. I think that fixes and mitigations for security are a good example for the non-hard rules mentioned in the [Contributing page](https://iceberg.apache.org/contribute/#contributing).

The improvement proposal document assumes 6 months migration period (aka two non-patch Iceberg/Java releases), which is IMO okay considering that this effort is about security.

[the remark this answers is not in the document]
Comment (Fokko) [reply open-api/rest-catalog-open-api.yaml:141]: I think Daniel is pointing at when the endpoint will go:

image

So that should be 1.8 or 2.0 according to the proposal, where it is cleaner to remove it at 2.0.

[the remark this answers is not in the document]
Comment (danielcweeks) [reply open-api/rest-catalog-open-api.yaml:141]: I think it's fine to include the warning in 1.6, but we shouldn't have a behavior change until 2.0 (I don't think a minor version release would be appropriate if we're significantly changing any of the behavior).  I would just update the notices to reference 2.0 for removal.

[the remark this answers is not in the document]
Comment (jbonofre) [reply open-api/rest-catalog-open-api.yaml:141]: I don't see blocker to have deprecation in 1.6.0 as soon as it doesn't actually change the current behavior (else it's a breaking change).
So I would deprecate (comment) for 1.6.0 and plan the removal in 1.8 or even better 2.0.

[the remark this answers is not in the document]
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:141]: To calm people down a bit: this change deprecates the endpoint, with the plan to remove the endpoint from the REST specification. That's nothing to do with the REST client implementation.

[the remark this answers is not in the document]
Comment (danielcweeks) [reply open-api/rest-catalog-open-api.yaml:141]: @snazy I don't really understand the pushback here.  When we deprecate something, we need to document the target for removal (this is part of our deprecation process).  While the client implementation is separate from the spec, the two are related in behavior.

To be clear, we need to include messaging on removal before this is ready to commit.

[the remark this answers is not in the document]
Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:141]: The misunderstanding is on my side. I've fixed it, but never pushed the change from my local branch.

File: open-api/rest-catalog-open-api.yaml [hunk 13]
```diff
@@ -134,9 +134,22 @@ paths:
...
+        implement this endpoint, unless you are fully aware of the potential security implications.
+
+        All clients are encouraged to explicitly set the configuration property `oauth2-server-uri`
+        to the correct OAuth endpoint.
+
+        Deprecated since Iceberg (Java) 1.6.0. The endpoint and related types will be removed from
+        this spec in Iceberg (Java) 1.7.0.
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:148]: what's the implication of removing this (and all the request / response types) from the Spec in 1.7.0 but not actually from the implementation? To me it seems that this should be marked for removal with Iceberg 2.0

Comment (snazy) [reply open-api/rest-catalog-open-api.yaml:148]: It's to let (new) adopters run into the trap of "blindly" implementing it and accidentally run into security issues

Comment (dimas-b) [reply open-api/rest-catalog-open-api.yaml:148]: Removing the endpoint from the OpenAPI YAML could break auto-generated clients. Custom clients (e.g. the Iceberg java REST client) calling this endpoint in servers that offer backward compatibility will not be affected. AFAIK, PyIceberg also _not_ affected by dropping the endpoint from OpenAPI.

Given the discussion of the negative security aspects of this endpoint (in the dev mail list), I tend to think that removing the endpoint from Open API sooner (1.7.0) is worth the potential hardship for auto-generated clients.

**Comment by ajantha-bhat** [comment]
I have added 1.6.0 milestone for this now because we still have a day or two for the release cut and we have mentioned that we are deprecating it in 1.6.0 in the PR changes.

File: open-api/rest-catalog-open-api.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (jbonofre) [reply open-api/rest-catalog-open-api.yaml:141]: @danielcweeks I'm confused. I think there's no problem to merge the deprecation flag and message right now, correct ? So, I propose to merge this PR right now, and plan "removal" (impacting client) on 2.0.

**Comment by snazy** [comment]
I've updated the PR to mention "2.0".

The CI failures look unrelated, but I don't have the power to rerun those.

**Comment by jackye1995** [comment]
Can you try rebase to see if it fixes the CI?

File: open-api/rest-catalog-open-api.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (danielcweeks) [reply open-api/rest-catalog-open-api.yaml:141]: I think we're good now.  The issue was that the target removal was 1.7 instead of 2.0, which I think was just confusion about deprecation vs. removal.

[the remark this answers is not in the document]
Comment (jbonofre) [reply open-api/rest-catalog-open-api.yaml:141]: @danielcweeks yes, that's why I was confused 😄 It seems we were "mixing" deprecation and removal. I agree it looks good now. Thanks !

File: open-api/rest-catalog-open-api.yaml [hunk 14]
```diff
@@ -134,9 +134,22 @@ paths:
...
       operationId: getToken
       description:
+        The `oauth/tokens` endpoint is **DEPRECATED for REMOVAL**. It is _not_ recommended to
+        implement this endpoint, unless you are fully aware of the potential security implications.
+
+        All clients are encouraged to explicitly set the configuration property `oauth2-server-uri`
+        to the correct OAuth endpoint.
+
+        Deprecated since Iceberg (Java) 1.6.0. The endpoint and related types will be removed from
+        this spec in Iceberg (Java) 2.0, whichever is released sooner.
```
Comment (danielcweeks) [inline open-api/rest-catalog-open-api.yaml:148]:
```suggestion
        this spec in Iceberg (Java) 2.0.
```

**Comment by snazy** [comment]
> Can you try rebase to see if it fixes the CI?

CI looking good

**Review by Fokko — APPROVED** [review]

**Comment by Fokko** [comment]
The vote has been passed: https://lists.apache.org/thread/o4qmrm5jx50mk1mqws0t9f1z2op4gvvm

**Review by nastra — APPROVED** [review]

**Comment by Fokko** [comment]
Thanks everyone, moving this forward for the 1.6.0 release 👍
